### PR TITLE
UI polishing - message layout animation

### PR DESCRIPTION
**Pull request by skorepak**
## In this PR, I have 
- fixed position of the reactions bubble based on figma [link](https://www.figma.com/file/ekifwChR9tR7zRJg1QEzSM/Chat-Sample-App?node-id=13285%3A230)
old:
Bez názvu 2
new:
Bez názvu
- added LayoutOption dataSource to the CollectionView layout to read the layout options and make diff
- added animation for appearing a disappearing reactions on the message

Video (first is standard speed, then again with Slow animation option in the simulator):
https://user-images.githubusercontent.com/7560049/121557894-6b677b80-ca15-11eb-895e-b8b930a44d33.mp4

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### ✅ Added
 - Add `staysConnectedInBackground` flag to `ChatClientConfig` (#1170)[https://github.com/GetStream/stream-chat-swift/pull/1170] 
 - Add `asView` helper for getting SwiftUI views from StreamChatUI UIViewControllers (#1174)[https://github.com/GetStream/stream-chat-swift/pull/1174]
```
Comment (skorepak) [inline CHANGELOG.md:13]: Seems to me that those links are broken 🤔  (swapped the bracket types)

File: Sources/StreamChatUI/ChatMessageList/StreamLayoutAttributes.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import UIKit
+
+class StreamLayoutAttributes: UICollectionViewLayoutAttributes {
+    var layoutOptions: ChatMessageLayoutOptions?
+    var appearingOptions: ChatMessageLayoutOptions?
+    var disappearingOptions: ChatMessageLayoutOptions?
```
Comment (evsaev) [inline Sources/StreamChatUI/ChatMessageList/StreamLayoutAttributes.swift:10]: should we make it `open`?

File: Sources/StreamChatUI/ChatMessageList/StreamLayoutAttributes.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        copy.disappearingOptions = disappearingOptions
+        return copy
+    }
+
+    override func isEqual(_ object: Any?) -> Bool {
+        if let customAttributes = object as? StreamLayoutAttributes {
+            if layoutOptions != customAttributes.layoutOptions
+                || appearingOptions != appearingOptions
+                || disappearingOptions != disappearingOptions
+            { return false }
```
Comment (evsaev) [inline Sources/StreamChatUI/ChatMessageList/StreamLayoutAttributes.swift:25]: When comparing `appearing/disappearing` options both are taken from the lhs 🤔 

I'd also switch to `guard`:
```swift
guard 
 let customAttributes = object as? StreamLayoutAttributes,
 layoutOptions == customAttributes.layoutOptions,
 appearingOptions == customAttributes.appearingOptions,
 disappearingOptions == customAttributes.disappearingOptions
 else { return false }

return super.isEqual(object)

```

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift [hunk 4]
```diff
@@ -34,7 +34,7 @@ open class _ChatMessageContentView<ExtraData: ExtraDataTypes>: _View, ThemeProvi
     /// The current layout options of the view.
     /// When this value is set the subviews are instantiated and laid out just once based on
     /// the received options.
-    private var layoutOptions: ChatMessageLayoutOptions!
+    private(set) var layoutOptions: ChatMessageLayoutOptions!
```
Comment (evsaev) [inline Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:37]: Don't remember why we decided to keep this private 🤔 and don't see any cons of giving public `read` access for this (but with moving away from force-unwrapped):
```swift
public private(set) var layoutOptions: ChatMessageLayoutOptions?
```

File: Sources/StreamChatUI/ChatMessageList/ChatMessageListCollectionViewLayout.swift [hunk 5]
```diff
@@ -89,6 +105,8 @@ open class ChatMessageListCollectionViewLayout: UICollectionViewLayout {
     /// Flag to make sure the `prepare()` function is only executed when the collection view had been loaded.
     /// The rest of the updates should come from `prepare(forCollectionViewUpdates:)`.
     private var didPerformInitialLayout = false
+    
+    public var dataSource: LayoutOptionsDataSource?
```
Comment (evsaev) [inline Sources/StreamChatUI/ChatMessageList/ChatMessageListCollectionViewLayout.swift:109]: has to be `weak` 🤔

**Review by evsaev — COMMENTED** [review]
this looks very good 👍 Let's fix the issues with datasource/checking attributes for being equal and 🚢 this

**Comment by VojtaStavik** [comment]
I'll finish this PR since it has some conflicts with the scrolling fixes I'm working on. Also, it seems the animation works well only for messages aligned to the left :) 


https://user-images.githubusercontent.com/6388510/121670887-378c6480-caae-11eb-9cf0-0fb3a51f02b9.mov

**Comment by VojtaStavik** [comment]
Closing this PR as the changes will be part of a bigger PR with scrolling improvements

**Comment by skorepak** [comment]
@VojtaStavik Can I delete the branch?

**Comment by VojtaStavik** [comment]
Yes, it's already included in my branch 👍

**Pull request closed without merging**